### PR TITLE
Fix to issue #696

### PR DESCRIPTION
**Pull request by kappeld**
This is a fix to issue #696. The connection function of the ConnectionManager has been updated to correctly handle MUSIC proxy nodes. A regression test has been added.

### CODE REVIEW

**Comment by mhoff** [comment]
The newly added testcase is apparently failing when ran with `make test`. However, on my local machine all mpi-related tests fail as I obviously have not properly configured `mpirun` in `.nestrc`.

@heplesser @jougs Could you elaborate on how to properly define `mpirun` in `.nestrc` for `mpirun (Open MPI) 1.6.5`? What did **not work** is simply uncommenting the existing lines:

```
/mpirun
[/integertype /stringtype]
[/numproc     /slifile]
{
 () [
  (mpirun -np ) numproc cvs ( ) statusdict/prefix :: (/bin/nest )  slifile
 ] {join} Fold
} Function def
```

When running `make test` this yields:

```
Running  /home/hoff/.local/bin/sli /home/hoff/.local/share/doc/nest/mpi_selftests/fail/crash_distributed_rank_invariant_collect_assert_or_die.sli
SLI v2.12.0 (C) 2004 The NEST Initiative
N_MPI: 1
Apr 12 17:46:43 mpirun [Error]: ArgumentType
    The type of the first parameter did not match the argument(s) of this function.
    Candidates for +mpirun+ are:
    integertype    stringtype     calls <proceduretype>
<end of output>
Test time =   0.12 sec
```
It might also be useful to make this more transparent for all users (?).

**Comment by heplesser** [comment]
@mhoff The call signature of the SLI `mpirun` command has changed. If you have an old `.nestrc` file in you home directory, it will not be overwritten or updated automatically, you either need to fix it by hand or delete the old `.nestc` file and the run NEST to install a new one. The `mpirun` section should look like this (three instead of two arguments):
```
% The command mpirun tells NEST how to assemble a string, which
% represents a valid command for calling a MPI-enabled executable in a
% distributed fashion The details of this string depend on your MPI
% library and the desired communication interface. Please modify the
% command according to your needs.

/mpirun
[/integertype /stringtype /stringtype]
[/numproc     /executable /scriptfile]
{
 () [
  (mpirun -np ) numproc cvs ( ) executable ( ) scriptfile
 ] {join} Fold
} Function def
```

**Comment by mhoff** [comment]
I think I see the problem now. I have used the pymusic interface to write the counterpart of the nest-node. But the testsuite does not support python-based nodes in the MUSIC tests.

We could change this by supporting python-based nodes, e.g.

```
diff --git a/testsuite/do_tests.sh.in b/testsuite/do_tests.sh.in
index 5c0e2b5..6b8c9f4 100755
--- a/testsuite/do_tests.sh.in
+++ b/testsuite/do_tests.sh.in
@@ -594,6 +594,10 @@ if test "x$(sli -c 'statusdict/have_music :: =')" = xtrue ; then
         sli_files=$(grep '\.sli' ${music_file} | sed -e "s#args=#${TESTDIR}#g")
         sli_files=$(for f in ${sli_files}; do if test -f ${f}; then echo ${f}; fi; done)
 
+        # Collect the list of python files from the .music file.
+        py_files=$(grep '\.py' ${music_file} | sed -e "s#args=#${TESTDIR}#g")
+        py_files=$(for f in ${py_files}; do if test -f ${f}; then echo ${f}; fi; done)
+
         # Check if there is an accompanying shell script for the test.
         sh_file=${TESTDIR}/$(basename ${music_file} .music).sh
         if test ! -f ${sh_file}; then unset sh_file; fi
@@ -608,7 +612,7 @@ if test "x$(sli -c 'statusdict/have_music :: =')" = xtrue ; then
         printf '%s' "  Running test '${test_name}' with $np $proc_txt... "
 
         # Copy everything to the tmpdir.
-        cp ${music_file} ${sh_file} ${sli_files} ${tmpdir}
+        cp ${music_file} ${sh_file} ${sli_files} ${py_files} ${tmpdir}
         cd ${tmpdir}
 
         # Create the runner script
```

or we can rewrite this test to not use python-based nodes.

Which solution is preferred?

**Comment by jougs** [comment]
@mhoff: I would prefer if the test was written in pure SLI as this guarantees that the tests also run in situations in which Python is not available. Feel free to ask if you require assistance with this.

I'm currently out of office, so please don't expect an immediate response. I'll do my best to still answer quickly, though.

**Comment by mhoff** [comment]
I understand.

The test is functional now. I reduced the receiver to a very simple sli-based node. The functionality is now reduced to setting up the connection and sending some spikes. The only assertion is that the program does not crash. This should suffice as a primitive regression test for the mapping problem.

File: testsuite/musictests/test_event_proxies_issue-696_receiver.sli [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+statusdict/have_music :: not { exit_test_gracefully } if
+
+M_ERROR setverbosity
+
+11 /N_NEURONS Set
+
+/subnet Create /NG Set
+NG ChangeSubnet
+/iaf_neuron N_NEURONS Create pop
```
Comment (stinebuu) [inline testsuite/musictests/test_event_proxies_issue-696_receiver.sli]: As iaf_neuron will soon be removed from NEST, could you change this to /iaf_psc_alpha ?

**Comment by mhoff** [comment]
@jougs Are you satisfied with the changes?

**Comment by jougs** [comment]
@heplesser: can you please pull master into your subnet-removal branch and remove subnet according to current standards once this is merged? I'll pull from there then. Thanks!

**Pull request merged**